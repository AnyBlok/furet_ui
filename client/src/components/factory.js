import Vue from 'vue';

const components = {};
const componentNames = [];

export const defineComponent = (elementName, declaration) => {
  let klass;
  if (elementName in components) {
    klass = components[elementName];
  } else {
    klass = {
      template: undefined,
      extend: [],
      prototypes: [],
    };
    components[elementName] = klass;
    componentNames.push(elementName);
  }
  if (declaration.template !== undefined) klass.template = declaration.template;
  if (declaration.prototype !== undefined) {
    klass.prototypes.push(declaration.prototype);
  }
  if (declaration.extend !== undefined) {
    declaration.extend.forEach((extend) => {
      klass.extend.push(extend);
    });
  }
};

window.defineComponent = defineComponent;

const getPrototypeFor = (elementName) => {
  const component = components[elementName];
  const prototypes = [];
  component.extend.forEach((otherElementName) => {
    prototypes.push(...getPrototypeFor(otherElementName));
  });
  prototypes.push(...component.prototypes);
  return prototypes;
};

export const createComponents = () => {
  componentNames.forEach((elementName) => {
    const component = components[elementName];
    const mixins = getPrototypeFor(elementName);
    Vue.component(elementName, (resolve) => {
      resolve({
        template: component.template,
        mixins,
      });
    });
  });
};
