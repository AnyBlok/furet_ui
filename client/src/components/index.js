import './mixins';
import './app';
import './homepage';
import './login';
import './logout';
import './about';
import './list';

export { createComponents, default } from './factory';
