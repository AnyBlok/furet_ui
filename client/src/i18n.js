import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

export const i18nConf = {
  locale: 'en',
  fallbackLocale: 'en',
  messages: {
    en: {
    },
  },
};
export const i18n = new VueI18n(i18nConf);
