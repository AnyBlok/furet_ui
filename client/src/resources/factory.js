import _ from 'underscore';

const resources = {};
export const resourceNames = [];

export const removeResource = (resourceName) => {
  const index = resourceNames.indexOf(resourceName);
  if (index !== -1) {
    delete resources[resourceName];
    resourceNames.splice(index, 1);
  }
};

export const defineResource = (resourceName, declaration) => {
  let klass;
  if (resourceName in resources) {
    klass = resources[resourceName];
  } else {
    klass = {
      children: [],
      name: resourceName,
      path: resourceName,
      templateName: resourceName,
      props: [],
      mustBeAuthenticated: false,
    };
    resources[resourceName] = klass;
    resourceNames.push(resourceName);
  }
  if (declaration.path !== undefined) klass.path = declaration.path;
  if (declaration.templateName !== undefined) klass.templateName = declaration.templateName;
  if (declaration.props !== undefined) klass.props = declaration.props;
  if (declaration.mustBeAuthenticated !== undefined) {
    klass.mustBeAuthenticated = declaration.mustBeAuthenticated;
  }
  if (declaration.children !== undefined) {
    declaration.children.forEach((child) => {
      let childDefinition = _.find(klass.children, c => c.name === child.name);
      if (childDefinition === undefined) {
        childDefinition = {
          name: child.name,
          completeName: `${resourceName}_${child.name}`,
          templateName: child.name,
          path: child.name,
          props: [],
          mustBeAuthenticated: false,
        };
        klass.children.push(childDefinition);
      }
      if (child.path !== undefined) childDefinition.path = child.path;
      if (child.templateName !== undefined) childDefinition.templateName = child.templateName;
      if (child.props !== undefined) childDefinition.props.push(...child.props);
      if (child.mustBeAuthenticated !== undefined) {
        childDefinition.mustBeAuthenticated = child.mustBeAuthenticated;
      }
    });
  }
};

window.defineResource = defineResource;

const createRoute = (resource, defaultTemplateProps) => {
  let path = `/${resource.path}`;
  let templateProps = defaultTemplateProps || '';
  if (resource.path === resource.name) {
    if (resource.props.length !== 0) {
      resource.props.forEach((prop) => {
        path += `/:${prop}`;
        templateProps += ` v-bind:${prop}="${prop}"`;
      });
    }
  } else if (resource.props.length !== 0) {
    resource.props.forEach((prop) => {
      templateProps += ` v-bind:${prop}="${prop}"`;
    });
  }
  const route = {
    name: resource.completeName || resource.name,
    path,
    props: templateProps !== '',
    meta: { requiresAuth: resource.mustBeAuthenticated },
    component: { props: resource.props },
  };
  if (resource.children === undefined || resource.children.length === 0) {
    route.component.template = `<${resource.templateName}${templateProps} />`;
  } else {
    route.children = [];
    if (resource.templateName === resource.name) {
      route.component.template = '<router-view />';
    } else {
      route.component.template = `<${resource.templateName}${templateProps} />`;
    }
    resource.children.forEach((child) => {
      route.children.push(createRoute(child, templateProps));
    });
  }
  return route;
};

export const getRoutes = () => {
  const routes = [];
  resourceNames.forEach((resourceName) => {
    const resource = resources[resourceName];
    routes.push(createRoute(resource));
  });
  return routes;
};
