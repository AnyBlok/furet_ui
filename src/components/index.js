import './mixins';
import './app';
import './homepage';
import './login';
import './logout';
import './pages';
import './fields';

export { createComponents, default } from './factory';
