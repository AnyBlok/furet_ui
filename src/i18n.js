/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import VueI18n from 'vue-i18n';
Vue.use(VueI18n);


export const i18nConf = {
    locale: 'en',
    fallbackLocale: 'en',
    messages: {
        en: {
            menus: {
                close: 'Close',
                search: 'Search ...',
            },
            views: {
                unknown: {
                    title: 'The wanted view "{name}" is unknown',
                    message: "Report this message to the administrator",
                },
                common: {
                    create: 'Create',
                    save: 'Save',
                    edit: 'Edit',
                    cancel: 'Cancel',
                    delete: 'Delete',
                    close: 'Close',
                    actions: 'Actions',
                    more: 'More',
                },
                clients: {
                    login: {
                        button: 'Log in',
                    },
                },
            },
            fields: {
                common: {
                    required: 'this field is required',
                },
                json: {
                    invalid: 'Invalid JSON format',
                },
                file: {
                    upload: 'Upload',
                    download: 'Download',
                    delete: 'Delete',
                },
            },
        },
    },
};
export const i18n = new VueI18n(i18nConf);
