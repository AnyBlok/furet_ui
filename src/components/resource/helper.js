import _ from 'underscore';
import { defineComponent } from '../factory';
import { safe_eval } from '../fields/common';

defineComponent('furet-ui-helper-mixin', {
  prototype: {
    props: ['resource', 'data', 'config'],
    inject: ['partIsReadonly'],
    computed: {
      isReadonly () {
        return this.getIsReadonly()
      },
      isHidden () {
        if (this.config.hidden == undefined) return false;
        return safe_eval(this.config.hidden, this.data || {}, this.resource);
      },
    },
    methods: {
      getIsReadonly () {
        if (this.resource.readonly) return true;
        if (this.partIsReadonly()) return true;
        const readonlyParams = safe_eval(this.config.readonly, this.data || {}, this.resource);
        if (this.config.writable) {
          const writableParams = safe_eval(this.config.writable, this.data || {}, this.resource);
          return readonlyParams && ! writableParams
        }
        return readonlyParams;
      },
    },
    provide: function () {
      return {
        partIsReadonly: this.getIsReadonly
      }
    },
  },
})

defineComponent('furet-ui-fieldset', {
  template: `
    <fieldset v-if="!isHidden" v-bind="config.props">
      <slot />
    </fieldset>
  `,
  extend: ['furet-ui-helper-mixin'],
})

defineComponent('furet-ui-tabs', {
  template: `
    <b-tabs 
      v-if="!isHidden" 
      ref="tabs" 
      v-model="resource.tabs[config.name]" 
      v-bind="config.props"
    >
      <slot />
    </b-tabs>
  `,
  extend: ['furet-ui-helper-mixin'],
})

const eval_counter = (value, fields) => {
  if (_.keys(fields).length === 0) return 0
  fields
  if (!value) return 0;
  let res = 0;
  try {
      res = eval(value);
  } catch (e) {
    console.log(e)
  }
  return res;
}

defineComponent('furet-ui-tab', {
  prototype: {
    functional: true,
    render: function (createElement, context) {
      const visible = (() => {
        if (context.props.config.hidden == undefined) return true;
        return !safe_eval(context.props.config.hidden, context.props.data || {}, context.props.resource);
      })();
      const options = Object.assign({}, context.data)
      options.attrs.visible= visible
      const props = context.props
      if (props.counter !== undefined) {
        const counter = eval_counter(props.counter, context.props.data);
        options.attrs.label = `${props.label} (${counter})`
      }
      return createElement( 'b-tab-item', options, [
        createElement('furet-ui-div', {props}, context.children)
      ])
    },
  },
})

defineComponent('furet-ui-div', {
  template: `
    <div v-if="!isHidden" v-bind="config.props">
      <slot />
    </div>
  `,
  extend: ['furet-ui-helper-mixin'],
  prototype: {
    computed: {
      isHidden () {
        if (this.config.hidden == undefined) return false;
        return safe_eval(this.config.hidden, this.data || {}, this.resource);
      },
    },
  },
})

defineComponent('furet-ui-selector', {
  template: `
    <furet-ui-form-field-common-tooltip-field
      v-bind:resource="resource"
      v-bind:data="data"
      v-bind:config="config"
    >
      <div class="field has-addons">
        <b-select 
          v-bind:disabled="isReadonly" 
          icon-pack="fa"
          v-bind:icon="config.icon"
          v-model="resource.selectors[config.name]"
          expanded
        >
          <option 
            v-for="option in getSelections"
            v-bind:key="option.value"
            v-bind:value="option.value"
          >
             {{ option.label }}
          </option>
        </b-select>
      </div>
    </furet-ui-form-field-common-tooltip-field>
  `,
  prototype: {
    props: ['resource', 'data', 'config'],
    computed: {
      value () {
        return this.resource.selectors[this.config.name];
      },
      isHidden () {
        if (this.config.hidden == undefined) return false;
        return safe_eval(this.config.hidden, this.data || {}, this.resource.selectors);
      },
      isReadonly () {
        const readonlyParams = safe_eval(this.config.readonly, this.data || {}, this.resource.selectors);
        return readonlyParams;
      },
      getSelections () {
        const colors = this.config.colors || {};
        const res = []
        res.push({label: '', value: null, color: null})
        _.each(this.config.selections, (label, value) => {
          res.push({value, label, color: colors[value]})
        });
        return res
      },
    },
    methods: {
      updateValue (value) {
        this.resource.selectors[this.config.name] = value;
      },
    },
  },
})
