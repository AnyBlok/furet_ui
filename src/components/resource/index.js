import './manager';
import './helper';
import './custom';
import './set';
import './list';
import './form';
