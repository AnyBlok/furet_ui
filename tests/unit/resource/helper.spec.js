import { mount } from "@vue/test-utils";
import { getComponentPrototype } from '@/components/factory';
import { i18n } from "@/i18n";
import { eval_counter } from "@/components/resource/helper";

const localVue = global.localVue;
const router = global.router;

const mocki18n = jest.spyOn(i18n, "_t");
mocki18n.mockResolvedValue('');

describe('furet-ui-fieldset component', () => {
  it('snapshot', () => {
    const wrapper = mount(getComponentPrototype('furet-ui-fieldset'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {},
        data: {},
        config: {},
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot readonly 1', () => {
    const wrapper = mount(getComponentPrototype('furet-ui-fieldset'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {},
        data: {},
        config: {},
      },
      provide: {
        partIsReadonly: () => true,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot readonly 2', () => {
    const wrapper = mount(getComponentPrototype('furet-ui-fieldset'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {},
        data: {},
        config: {
          readonly: true
        },
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot: hidden', () => {
    const wrapper = mount(getComponentPrototype('furet-ui-fieldset'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {},
        data: {},
        config: {
          hidden: true
        },
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});

describe('furet-ui-tabs/furet-ui-tab component', () => {
  it('snapshot', () => {
    const tab = getComponentPrototype('furet-ui-tab')
    const wrapper = mount(getComponentPrototype('furet-ui-tabs'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {
          tabs: {}
        },
        data: {},
        config: {
          name: 'tabs',
        },
      },
      slots: {
        default: ['<furet-ui-tab :resource="{}" :data="{}" :config="{}" />']
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot counter', () => {
    const tab = getComponentPrototype('furet-ui-tab')
    const wrapper = mount(getComponentPrototype('furet-ui-tabs'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {
          tabs: {}
        },
        data: {},
        config: {
          name: 'tabs',
        },
      },
      slots: {
        default: ['<furet-ui-tab :resource="{}" :data="{}" :config="{}" counter="1"/>']
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot hidden', () => {
    const tab = getComponentPrototype('furet-ui-tab')
    const wrapper = mount(getComponentPrototype('furet-ui-tabs'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {
          tabs: {}
        },
        data: {},
        config: {
          name: 'tabs',
        },
      },
      slots: {
          default: ['<furet-ui-tab :resource="{}" :data="{}" :config="{hidden: true}" />']
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('eval_counter', () => {
    expect(eval_counter(1, {})).toBe(0)
  });
  it('eval_counter', () => {
    expect(eval_counter(undefined, {field: 'test'})).toBe(0)
  });
  it('eval_counter 1', () => {
    expect(eval_counter('fields.field', {field: 1})).toBe(1)
  });
  it('eval_counter 2', () => {
    expect(eval_counter('fields.field', {field: [1, 2]})).toBe(2)
  });
  it('eval_counter 3', () => {
    expect(eval_counter('fields.field', {field: {count: 3}})).toBe(3)
  });
});

describe('furet-ui-div component', () => {
  it('snapshot', () => {
    const wrapper = mount(getComponentPrototype('furet-ui-div'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {},
        data: {},
        config: {},
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('is readonly 1', () => {
    const wrapper = mount(getComponentPrototype('furet-ui-div'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {},
        data: {},
        config: {},
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.vm.isReadonly).toBe(false)
  });
  it('is readonly 2', () => {
    const wrapper = mount(getComponentPrototype('furet-ui-div'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {},
        data: {},
        config: {
          readonly: false,
          writable: true
        },
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.vm.isReadonly).toBe(false)
  });
  it('snapshot: hidden', () => {
    const wrapper = mount(getComponentPrototype('furet-ui-div'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {},
        data: {},
        config: {
          hidden: true,
        },
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});

describe('furet-ui-selector component', () => {
  it('snapshot', () => {
    const selectors = {}
    const wrapper = mount(getComponentPrototype('furet-ui-selector'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {
          selectors,
        },
        data: {},
        config: {
          name: 'selector',
          selections: {
            field1: 'Value 1',
            field2: 'Value 2',
            field3: 'Value 3',
          },
        },
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot: readonly by provide', () => {
    const selectors = {}
    const wrapper = mount(getComponentPrototype('furet-ui-selector'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {
          selectors,
        },
        data: {},
        config: {
          name: 'selector',
          selections: {
            field1: 'Value 1',
            field2: 'Value 2',
            field3: 'Value 3',
          },
        },
      },
      provide: {
        partIsReadonly: () => true,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot: readonly by config', () => {
    const selectors = {}
    const wrapper = mount(getComponentPrototype('furet-ui-selector'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {
          selectors,
        },
        data: {},
        config: {
          name: 'selector',
          readonly: true,
          selections: {
            field1: 'Value 1',
            field2: 'Value 2',
            field3: 'Value 3',
          },
        },
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot: hidden', () => {
    const selectors = {}
    const wrapper = mount(getComponentPrototype('furet-ui-selector'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {
          selectors,
        },
        data: {},
        config: {
          name: 'selector',
          hidden: true,
          selections: {
            field1: 'Value 1',
            field2: 'Value 2',
            field3: 'Value 3',
          },
        },
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });
  it('snapshot: updateValue', () => {
    const selectors = {}
    const wrapper = mount(getComponentPrototype('furet-ui-selector'), {
      localVue, 
      router,
      i18n,
      propsData: {
        resource: {
          selectors,
        },
        data: {},
        config: {
          name: 'selector',
          hidden: true,
          selections: {
            field1: 'Value 1',
            field2: 'Value 2',
            field3: 'Value 3',
          },
        },
      },
      provide: {
        partIsReadonly: () => false,
      }
    });
    wrapper.vm.updateValue('field1')
    expect(selectors).toStrictEqual({"selector": "field1"})
  });
});
