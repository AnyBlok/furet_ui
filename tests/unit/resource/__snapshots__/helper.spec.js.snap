// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`furet-ui-div component snapshot 1`] = `<div />`;

exports[`furet-ui-div component snapshot: hidden 1`] = `<!---->`;

exports[`furet-ui-fieldset component snapshot 1`] = `<fieldset />`;

exports[`furet-ui-fieldset component snapshot readonly 1 1`] = `<fieldset />`;

exports[`furet-ui-fieldset component snapshot readonly 2 1`] = `<fieldset />`;

exports[`furet-ui-fieldset component snapshot: hidden 1`] = `<!---->`;

exports[`furet-ui-selector component snapshot 1`] = `
<span
  class="is-primary is-top is-medium b-tooltip"
  data-label=""
  style="transition-delay: 0ms; width: 100%;"
>
  <div
    class="field"
    style="width: inherit;"
  >
    <!---->
      
    <!---->
  </div>
</span>
`;

exports[`furet-ui-selector component snapshot: hidden 1`] = `<div />`;

exports[`furet-ui-selector component snapshot: readonly by config 1`] = `
<span
  class="is-primary is-top is-medium b-tooltip"
  data-label=""
  style="transition-delay: 0ms; width: 100%;"
>
  <div
    class="field"
    style="width: inherit;"
  >
    <!---->
      
    <!---->
  </div>
</span>
`;

exports[`furet-ui-selector component snapshot: readonly by provide 1`] = `
<span
  class="is-primary is-top is-medium b-tooltip"
  data-label=""
  style="transition-delay: 0ms; width: 100%;"
>
  <div
    class="field"
    style="width: inherit;"
  >
    <!---->
      
    <!---->
  </div>
</span>
`;

exports[`furet-ui-tabs/furet-ui-tab component snapshot 1`] = `
<div
  class="b-tabs"
>
  <nav
    class="tabs"
  >
    <ul />
  </nav>
   
  <section
    class="tab-content"
  >
    <div
      class="tab-item"
      config="[object Object]"
      data="[object Object]"
      resource="[object Object]"
      style="display: none;"
    >
      <div />
    </div>
  </section>
</div>
`;

exports[`furet-ui-tabs/furet-ui-tab component snapshot counter 1`] = `
<div
  class="b-tabs"
>
  <nav
    class="tabs"
  >
    <ul />
  </nav>
   
  <section
    class="tab-content"
  >
    <div
      class="tab-item"
      config="[object Object]"
      counter="1"
      data="[object Object]"
      resource="[object Object]"
      style="display: none;"
    >
      <div />
    </div>
  </section>
</div>
`;

exports[`furet-ui-tabs/furet-ui-tab component snapshot hidden 1`] = `
<div
  class="b-tabs"
>
  <nav
    class="tabs"
  >
    <ul />
  </nav>
   
  <section
    class="tab-content"
  >
    <div
      class="tab-item"
      config="[object Object]"
      data="[object Object]"
      resource="[object Object]"
      style="display: none;"
    >
      <!---->
    </div>
  </section>
</div>
`;
